rust/src/measurements/gumbel_max/mod.rs

```diff
@@ -63,43 +89,76 @@ where
-                })
-                .reduce(|l, r| {
-                    let (mut l, mut r) = (l?, r?);
-                    Ok(if l.1.greater_than(&mut r.1)? { l } else { r })
-                })
-                .ok_or_else(|| err!(FailedFunction, "there must be at least one candidate"))?
-                .map(|v| v.0)
+            select_score(arg.iter().cloned(), optimize.clone(), scale_frac.clone())
         }),
         input_metric.clone(),
         MaxDivergence::default(),
-        PrivacyMap::new_fallible(move |d_in: &TIA| {
-            // convert L_\infty distance to range distance
-            let d_in = input_metric.range_distance(*d_in)?;
+        PrivacyMap::new_fallible(rnm_gumbel_map(scale, input_metric)),
+    )
+}
+
+pub(crate) fn rnm_gumbel_map<QI, QO>(
+    scale: QO,
+    input_metric: LInfDistance<QI>,
+) -> impl Fn(&QI) -> Fallible<QO>
+where
+    QI: Clone + InfAdd,
+    QO: Float + InfCast<QI>,
+{
+    move |d_in: &QI| {
+        // convert L_\infty distance to range distance
+        let d_in = input_metric.range_distance(d_in.clone())?;
 
-            // convert data type to QO
-            let d_in = QO::inf_cast(d_in)?;
+        // convert data type to QO
+        let d_in = QO::inf_cast(d_in)?;
 
-            if d_in.is_sign_negative() {
-                return fallible!(InvalidDistance, "sensitivity must be non-negative");
-            }
+        if d_in.is_sign_negative() {
+            return fallible!(InvalidDistance, "sensitivity must be non-negative");
+        }
 
-            if scale.is_zero() {
-                return Ok(QO::infinity());
-            }
+        if scale.is_zero() {
+            return Ok(QO::infinity());
+        }
 
-            // d_out >= d_in / scale
-            d_in.inf_div(&scale)
-        }),
-    )
+        // d_out >= d_in / scale
+        d_in.inf_div(&scale)
+    }
+}
+
+pub fn select_score<TIA>(
+    iter: impl Iterator<Item = TIA>,
+    optimize: Optimize,
+    scale: RBig,
+) -> Fallible<usize>
+where
+    TIA: Number + CastInternalRational,
+{
+    if scale.is_zero() {
+        let cmp = |l: &TIA, r: &TIA| match optimize {
+            Optimize::Max => l > r,
+            Optimize::Min => l < r,
+        };
+        return Ok(iter
+            .enumerate()
+            .reduce(|l, r| if cmp(&l.1, &r.1) { l } else { r })
+            .ok_or_else(|| err!(FailedFunction, "there must be at least one candidate"))?
+            .0);
+    }
+
+    (iter.enumerate())
+        .map(|(i, v)| {
+            let mut shift = v.into_rational()?;
+            if optimize == Optimize::Min {
+                shift = -shift;
+            }
+            Ok((i, GumbelPSRN::new(shift, scale.clone())))
+        })
+        .reduce(|l, r| {
+            let (mut l, mut r) = (l?, r?);
+            Ok(if l.1.greater_than(&mut r.1)? { l } else { r })
+        })
+        .ok_or_else(|| err!(FailedFunction, "there must be at least one candidate"))?
+        .map(|v| v.0)
```

Comment:
Do we have a test anywhere of `max` vs `min`? I could imagine getting the logic backwards by mistake.

Reply:
Added one!